framework/src/modules/chain/submodules/blocks/process.js

```diff
@@ -427,11 +435,242 @@ class Process {
 			processTransactions: scope.modules.processTransactions,
 		};
 
+		deleteLastBlock = promisify(modules.blocks.chain.deleteLastBlock);
+
 		// Set module as loaded
 		__private.loaded = true;
 	}
+
+	/**
+	 * Handle newly received block based on the new fork choice rule
+	 *
+	 * @listens module:transport~event:receiveBlock
+	 * @param {block} block - New block
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_forkChoice(block) {
+		// Current slot number based on current time since LiskEpoch ~ Slot number at the time the new block is received
+		// Better to do it here rather than in the Sequence so reciving time is more accurate
+		const newBlockReceivedAt = slots.getTime();
+
+		// Execute in sequence via sequence
+		return library.sequence.add(callback => {
+			this._forkChoiceTask(block, newBlockReceivedAt)
+				.then(result => callback(null, result))
+				.catch(error => callback(error));
+		});
+	}
+
+	/**
+	 * Wrap of fork choice rule logic so it can be added to Sequence and properly tested
+	 * @param block
+	 * @param newBlockReceivedAt - Time when the block was received
+	 * @return {Promise}
+	 * @private
+	 */
+	async _forkChoiceTask(block, newBlockReceivedAt) {
+		const lastBlock = modules.blocks.lastBlock.get();
+
+		const forgingSlotLastBlock = slots.getSlotNumber(lastBlock.timestamp);
+		const forgingSlotNewBlock = slots.getSlotNumber(block.timestamp);
+
+		// Slot number when lastBlock was received.
+		const lastBlockReceivedAt =
+			modules.blocks.lastReceipt.get() || lastBlock.timestamp;
+
+		if (lastBlock.id === block.id) {
+			// Case 1: same block received twice
+			return this._handleSameBlockReceived(block);
+		}
+
+		if (
+			lastBlock.height + 1 === block.height &&
+			block.previousBlock === lastBlock.id
+		) {
+			// Case 2: correct block received
+			return this._handleGoodBlock(block);
+		}
+
+		if (
+			lastBlock.height === block.height &&
+			lastBlock.heightPrevoted === block.heightPrevoted &&
```

Comment:
Please use the different names for these attributes as we discussed earlier in a meeting. https://github.com/LiskHQ/lisk-sdk/blob/64488f43f7e56d6b963780aabd0f034c6eeb8895/framework/test/mocha/fixtures/blocks.js#L77-L84

Reply:
Good catch, done.